repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -0,0 +1,360 @@
+                  } else {
+                    Map.empty[String, String]
+                  }
+
+                  val tenantName = OpenStackServiceHeader.TENANT_NAME.toString -> validToken.tenantName
+
+                  val defaultRegion = validToken.defaultRegion.map(dr => OpenStackServiceHeader.DEFAULT_REGION.toString -> dr)
+
+                  val contactId = validToken.contactId.map(ci => OpenStackServiceHeader.CONTACT_ID.toString -> ci)
+
+                  val expirationDate = OpenStackServiceHeader.X_EXPIRATION.toString -> validToken.expirationDate
+
+                  val impersonatorId = validToken.impersonatorId.map(iid => OpenStackServiceHeader.IMPERSONATOR_ID.toString -> iid)
+                  val impersonatorName = validToken.impersonatorName.map(in => OpenStackServiceHeader.IMPERSONATOR_NAME.toString -> in)
+
+                  val identityStatus = OpenStackServiceHeader.IDENTITY_STATUS.toString -> IdentityStatus.Confirmed.toString
+
+                  Pass(headers ++ userHeaders ++ rolesHeader + tenantName + xAuthHeader ++ defaultRegion ++ contactId
+                    + expirationDate ++ impersonatorId ++ impersonatorName + identityStatus)
+                case reject: Reject => reject
+              }
+
+              addHeaders
+            case Failure(x: IdentityAdminTokenException) =>
+              Reject(SC_INTERNAL_SERVER_ERROR, failure = Some(x))
+            case Failure(x: IdentityCommuncationException) =>
+              Reject(SC_BAD_GATEWAY, failure = Some(x))
+            case Failure(x) =>
+              //TODO: this isn't yet complete, should return other status codes?
+              Reject(SC_INTERNAL_SERVER_ERROR, failure = Some(x))
+          }
+        } getOrElse {
+          //Don't have an auth token to validate
+          Reject(SC_FORBIDDEN, Some("Auth token not found in headers"))
+        }
+      }
+
+      //Handle the result of the filter
+      result match {
+        case rejection: Reject =>
+          val message: Option[String] = rejection match {
+            case Reject(_, Some(x), _) => Some(x)
+            case Reject(code, None, Some(failure)) => Some(failure.getMessage)
+            case _ => None
+          }
+          //Handle delegation if necessary
+          Option(config.getDelegating) match {
+            case Some(delegating) =>
+              logger.debug(s"Delegating with status ${rejection.status}")
+              val delegationHeaders = buildDelegationHeaders(rejection.status,
+                "keystone-v2",
+                message.getOrElse("Failure in the Keystone v2 filter").replace("\n", ""),
+                delegating.getQuality)
+
+              delegationHeaders foreach { case (key, values) =>
+                values foreach { value =>
+                  request.addHeader(key, value)
+                }
+              }
+              request.addHeader(OpenStackServiceHeader.IDENTITY_STATUS.toString, IdentityStatus.Indeterminate.toString)
+
```

Comment:
so this is what i was talking about above, in delegating mode we should be populating as many of the headers as we managed to make so herp can construct the best messages possible.